### PR TITLE
gridstuff

### PR DESCRIPTION
**Pull request by ocefpaf**
@phobson Do you want to be a maintainer of these recipes here at conda-forge?

### CODE REVIEW

**Comment by ocefpaf** [comment]
Not sure what is going on there. I will take a look tomorrow. Time to :sleeping:

**Comment by pelson** [comment]
```
===== testing package: pygridgen-0.1.1-py27_0 =====
import: 'pygridgen'
/opt/conda/envs/_test/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
running run_test.py
Traceback (most recent call last):
  File "/opt/conda/conda-bld/test-tmp_dir/run_test.py", line 37, in <module>
    grid = pygridgen.grid.Gridgen(x, y, beta, shape=(20, 20), focus=focus)
  File "/opt/conda/envs/_test/lib/python2.7/site-packages/pygridgen/grid.py", line 767, in __init__
    self._libgridgen.gridgen_generategrid2.restype = ctypes.POINTER(ctypes.c_void_p)
AttributeError: 'Gridgen' object has no attribute '_libgridgen'
Exception AttributeError: "'Gridgen' object has no attribute '_libgridgen'" in <bound method Gridgen.__del__ of <pygridgen.grid.Gridgen object at 0x7f6b868dc2d0>> ignored

TESTS FAILED: pygridgen-0.1.1-py27_0
```

File: recipes/pygridgen/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    build:
+        - python
+        - numpy
+    run:
+        - python
+        - pandas
+        - numpy
+        - matplotlib
+        - pyproj
+        - gridutils
```
Comment (phobson) [inline recipes/pygridgen/meta.yaml]: add `gridgen` here and that should fix the failure

**Comment by phobson** [comment]
@ocefpaf @pelson yes! I'd be happy too. I've never used CircleCI before but looking around, everything seems fairly intuitive

File: recipes/pygridgen/meta.yaml (hunk 1 above)
Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: Oops :flushed: I wanted to leave only the last on in the stack (`gridgen`) but ended but deleting the wrong one.

**Comment by ocefpaf** [comment]
> @ocefpaf @pelson yes! I'd be happy too. I've never used CircleCI before but looking around, everything seems fairly intuitive

As a recipe maintainer all we have to do is to update the software version when there is a new release. (And fix the recipe here and there sometimes.)

**Comment by ocefpaf** [comment]
@jakirkham or @pelson can you take a quick look and merge if things are alright?

**Comment by jakirkham** [comment]
One second, some of the feedstocks weren't created. Looking at that then we can merge things once that is cleaned up.

**Comment by ocefpaf** [comment]
I noticed that Travis-CI built `rank_filter`, which is not part of this PR. Not sure what is going on...

File: recipes/csa/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    git_url: https://github.com/sakov/csa-c.git
+    git_tag: v1.22.0
+
+build:
+    number: 0
+    skip: True  # [win]
+
+requirements:
+    run:
+        - libgcc
```
Comment (jakirkham) [inline recipes/csa/meta.yaml]: Need to think about this one a bit more.

**Comment by jakirkham** [comment]
So there are a few places where we are including `libgcc` as a runtime library. Though we aren't necessarily using conda's `gcc`. Should we be including `libgcc` as a runtime requirement if it might not be the version we linked against.

File: recipes/csa/build.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+pushd csa
+
+./configure --prefix=$PREFIX
+make
+make tests
+make install
+
+popd
```
Comment (jakirkham) [inline recipes/csa/build.sh]: Probably unnecessary, but it doesn't hurt anything.

File: recipes/gridgen/build.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+export LDFLAGS="-L$PREFIX/lib"
+export CPPFLAGS="-I/$PREFIX/include $CPPFLAGS"
+
+pushd gridgen
+
+./configure --prefix=$PREFIX
+
+make
+make lib
+make shlib
```
Comment (jakirkham) [inline recipes/gridgen/build.sh]: Interesting. So the library building isn't in `make all`? Also, I am guessing we are building static and shared libraries here. Is that correct?

File: recipes/pygridgen/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        - pyproj
+        - gridgen
+
+test:
+    imports:
+        - pygridgen
+
+about:
+    home: https://github.com/hetland/pygridgen.git
+    license:  BSD License
```
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: License version?

File: recipes/pygridgen/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    version: 0.1.1
+
+source:
+    git_url: https://github.com/phobson/pygridgen.git
+    git_tag: 0.1.1
+
+build:
+    number: 0
+    skip: True  # [win]
+    script: python setup.py install
```
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Do we need this ` --single-version-externally-managed --record record.txt`?

File: recipes/pygridtools/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package:
+    name: pygridtools
+    version: 0.2
+
+source:
+    git_url: https://github.com/Geosyntec/pygridtools.git
+    git_tag: v0.2
```
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: Could we download the file from the releases page? Also, some kind of checksum would be nice here.

File: recipes/pygridtools/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    imports:
+        - pygridtools
+    commands:
+        - nosetests
+    requires:
+        - nose
+
+about:
+    home: https://github.com/Geosyntec/pygridtools.git
+    license:  BSD License
```
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: License version?

File: recipes/pygridtools/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+test:
+    imports:
+        - pygridtools
+    commands:
+        - nosetests
+    requires:
+        - nose
+
+about:
+    home: https://github.com/Geosyntec/pygridtools.git
```
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: Probably can drop the `.git` here.

File: recipes/pygridgen/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        - matplotlib
+        - pyproj
+        - gridgen
+
+test:
+    imports:
+        - pygridgen
+
+about:
+    home: https://github.com/hetland/pygridgen.git
```
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Also, drop `.git` here.

**Comment by jakirkham** [comment]
There are few `home` URLs that end with `.git`. I marked a few, but not all. Could drop those extra `.git`s.

File: recipes/csa/meta.yaml (hunk 2 above)
Comment (ocefpaf) [inline recipes/csa/meta.yaml]: I am not 100% sure either. Adding `libgcc` worked out for me in `nco` and `cdo` to solve problems like [this one](https://github.com/ioos/conda-recipes/issues/723).

File: recipes/pygridgen/meta.yaml (hunk 5 above)
Comment (phobson) [inline recipes/pygridgen/meta.yaml]: Upstream (https://github.com/hetland/pygridgen) doesn't specify a license. He knows I've forked and made improvements and PRs (that have gone stale). He hasn't raised concerns about my use in private industry, FWIW.

Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: He updated that to MIT: https://github.com/hetland/pygridgen/blob/master/setup.py#L13

File: recipes/pygridtools/meta.yaml (hunk 8 above)
Comment (phobson) [inline recipes/pygridtools/meta.yaml]: I default to the 3-clause BSD. My goals is that these are released under a permissive license.

File: recipes/pygridgen/meta.yaml (hunk 5 above)
Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: Or maybe not https://github.com/hetland/pygridgen/blob/master/setup.py#L34 :unamused:

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (ocefpaf) [inline recipes/pygridtools/meta.yaml]: Not sure if that is necessary when using git. Git is pretty robust to ensure that we have the "correct" clone. (We must add checksum if using files though.)

File: recipes/pygridgen/meta.yaml (hunk 6 above)
Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: No `setuptools`. We cannot do that :unamused:

File: recipes/gridgen/build.sh (hunk 4 above)
Comment (ocefpaf) [inline recipes/gridgen/build.sh]: > So the library building isn't in `make all`

Last time I checked no. I have to re-visit upstream to see if there is any improvements.

> Also, I am guessing we are building static and shared libraries here. Is that correct?

Correct.

**Comment by ocefpaf** [comment]
> So there are a few places where we are including libgcc as a runtime library. Though we aren't necessarily using conda's gcc. Should we be including libgcc as a runtime requirement if it might not be the version we linked against.

I am not sure. That did helped me with `nco`. @msarahan is this a good or a bad practice?

File: recipes/pygridgen/meta.yaml (hunk 6 above)
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Totally fine. I have had a few cases like that too. Just wanted to confirm. Thanks.

File: recipes/gridgen/build.sh (hunk 4 above)
Comment (phobson) [inline recipes/gridgen/build.sh]: @jakirkham 

Yes (static and shared libs).
 
In the early days of me using these libraries, the incantations make everything link properly involved a little bit of magic, making sure the handles of the pots and pans in your kitchen were pointed north, and a little bit of luck. 

Unfortunately, `autoconf`, configure scripts, etc, are very much a black box to me. What you see above is the first procedure I found that reliable built everything. One it worked, I left it alone.

In short, this was me flailing around until everything worked.

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: Sure, I guess my thought is some git repos can become very huge and then this download step will take time (and some space) versus simply downloading a tarball of a release. After looking more closely, this repo does not seem to be one of those case. So, I leave it up to you.

File: recipes/gridgen/build.sh (hunk 4 above)
Comment (jakirkham) [inline recipes/gridgen/build.sh]: Alright, I trust you guys have spent enough time flailing around. :wink:
Just asking for my own edification. :smile:

File: recipes/csa/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/csa/meta.yaml]: Ack, yes, I am familiar with `libgomp` fun. Let me think about it a little more.

File: recipes/gridgen/build.sh (hunk 4 above)
Comment (ocefpaf) [inline recipes/gridgen/build.sh]: @phobson you took this farther than most of the previous users.  That is good enough to me!

File: recipes/pygridgen/meta.yaml (hunk 5 above)
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Initially, I was just meaning what BSD flavor, but I guess this is more complicated.

I opened this issue ( https://github.com/hetland/pygridgen/issues/27 ) to see if he will clarify. I am leaning towards saying MIT as it is referred to in a couple of places [here]( https://github.com/hetland/pygridgen/blob/9917ab351385a2311e286452815da3fe148b8234/setup.py#L13 ) and [here]( https://github.com/hetland/pygridgen/blob/9917ab351385a2311e286452815da3fe148b8234/pygridgen/__init__.py#L11 ). That [BSD]( https://github.com/hetland/pygridgen/blob/9917ab351385a2311e286452815da3fe148b8234/setup.py#L34 ) could have been leftover. Though we could say BSD or MIT and leave a note here to check on the issue above and update appropriately once clarified.

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (ocefpaf) [inline recipes/pygridtools/meta.yaml]: > some git repos can become very huge

You have a point there. Not sure how conda clones the repository (what is the depth of the clone?). I don't like to use GitHub tagged files though. Users can delete them and re-add a new one with the same tag but fixing a problem, and then you have a checksum issue on your hands. PyPI solved that by forcing people to bump the version because PyPI does not accept the same file name/version for the same release even if you delete it.

(It would be nice to have `pygridtools` source listed at PyPI and then I would use the file instead. @phobson interested?)

File: recipes/pygridgen/meta.yaml (hunk 5 above)
Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: > I opened this issue ( hetland/pygridgen#27 ) to see if he will clarify.

Thanks!

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: > Not sure how conda clones the repository (what is the depth of the clone?).

Default is to clone the whole history. There was an option to do shallow clones that was proposed. Should be in `conda build` 1.19.0 and later. Just set `git_depth` to the number of commits you would want.

> Users can delete them and re-add a new one with the same tag but fixing a problem, and then you have a checksum issue on your hands.

User could delete tags too. Generally, I feel these are bad practices, but you're right. We can always by bitten by them if someone does them anyways.

> PyPI solved that by forcing people to bump the version because PyPI does not accept the same file name/version for the same release even if you delete it.

Though that can get a bit extreme. Like if the upload got corrupted or PyPI had issues. I remember that happening to NumPy before.

<br>
In any event, given this is a relatively small repo. I am ok with it as is. If you want to try a shallow clone, feel free. If you guys work out something so it is on PyPI, I am good with that, as well. I leave it up to you.

Comment (phobson) [inline recipes/pygridtools/meta.yaml]: @ocefpaf I can upload to pip. but it'll probably be ~a week before i get to it.

File: recipes/pygridgen/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        - pyproj
+        - gridgen
+
+test:
+    imports:
+        - pygridgen
+
+about:
+    home: https://github.com/hetland/pygridgen
+    license:  MIT
```
Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Can we leave a comment here referencing this issue ( https://github.com/hetland/pygridgen/issues/27 ) so that we remember to check and make sure this is ok?

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (ocefpaf) [inline recipes/pygridtools/meta.yaml]: No rush and don't do that on this recipe account! Only if you think it is worth it.

File: recipes/pygridgen/meta.yaml (hunk 11 above)
Comment (ocefpaf) [inline recipes/pygridgen/meta.yaml]: Makes sense. (And done :wink:)

Comment (jakirkham) [inline recipes/pygridgen/meta.yaml]: Excellent, thanks.

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (phobson) [inline recipes/pygridtools/meta.yaml]: I'll need to mull that over a bit. Uploading to pip always feels like a hassle to me. But that's mostly b/c i don't do it enough.

**Comment by jakirkham** [comment]
Alright, think this is looking pretty good at this point. Really only need to think about the `libgomp` issue.

File: recipes/pygridtools/meta.yaml (hunk 7 above)
Comment (jakirkham) [inline recipes/pygridtools/meta.yaml]: Alright, I think we are all ok with this part. Let's freeze it. Sound good?

**Comment by ocefpaf** [comment]
> Really only need to think about the `libgomp` issue.

Let's wait until tomorrow for @msarahan and @pelson's opinions. (I usually just keep trying until things work. I am sure you guys will have a better understanding of what needs to be done to solve that properly. If not we can always create docs telling people to install the missing libs :wink:)

**Comment by jakirkham** [comment]
So, what I am debating at this point is whether we should just use `conda`'s `gcc` compiler in this case. Right now I am leaning towards yes.

Though this kind of goes against our "use system compilers" philosophy. However, that hasn't been a hard rule.

It just seems if we are bundling the `libgcc` from conda we should using the copy of `gcc` that should link against it.

I agree we should see what other people think.

**Comment by ocefpaf** [comment]
> So, what I am debating at this point is whether we should just use `conda`'s `gcc` compiler in this case. Right now I am leaning towards yes.

I was on that team until a few messages convinced me otherwise. We even have an opened issue for that: https://github.com/conda-forge/conda-forge.github.io/issues/29

**Comment by jakirkham** [comment]
There are some issues for Mac, as well. Not all versions of Mac have `gcc` and thus `libgomp` for instance. Eventually `clang` developed and merged in [their own OpenMP implementation]( https://clang-omp.github.io/ ). Though I am unclear what XCode versions this has actually shown up in (if any), but can guarantee it is not when `gcc` is removed.

**Comment by jakirkham** [comment]
I thought about just bundling the `libgcc` stuff from our docker image in some fashion, but I don't want to clash with conda's `libgcc`.

Though we could just try to bundle the missing pieces by having packages only for use in the VMs that copy those pieces into the conda environment. For instance, we could bundle `libgfortran`, `libgomp`, and others to solve these kinds of problems. It would be an only Mac and Linux solution, but it could be functional.

The only worry in these cases is if conda's `libgcc` gets installed it will overwrite these things and potentially introduce ABI incompatibilities.

**Comment by msarahan** [comment]
If you use my recipe for libgcc, you would be doing both (getting it from Docker, and also being the same as conda's libgcc.

https://github.com/conda-forge/staged-recipes/pull/135

**Comment by jakirkham** [comment]
How is this working in the Mac case though?

**Comment by msarahan** [comment]
It doesn't.  We use the system compiler there.

**Comment by jakirkham** [comment]
Meaning there is no Mac libgcc recipe for that case?

**Comment by msarahan** [comment]
I don't see how it would be necessary if nothing is compiled with it?

If it is necessary for any reason, then I'm happy to take that up internally - I just don't know how it interplays with the rest of the Mac build system

**Comment by jakirkham** [comment]
Same reasons as mentioned on the other issue GNU OpenMP and GFortran. In this case, only OpenMP is relevant.

**Comment by msarahan** [comment]
I see, thanks.  I don't know the landscape nearly as well as Linux.  I need to discuss this with @mingwandroid and @groutr.

Since there is no docker image, we may need gcc packages on Mac.  That's unfortunate - I'm guessing that will increase demand for them on Linux as well.

**Comment by jakirkham** [comment]
Maybe we can just create a recipe that selects one of [these]( https://gcc.gnu.org/wiki/GFortranBinaries ) and installs them. Though we might be able provide the 10.7 one for all cases. We have to change the install location. We could rip the libraries out into a package for use as a runtime dependency to keep things light and consistent. As the `gcc` compiler, it slowly gets removed on Mac, so we don't have to worry about conflicts. Also, Mac continued to use gcc 4.2.1 until it was removed. So, we are certain we always have a newer version of `gcc` than the system even when the system provides a version.

**Comment by jakirkham** [comment]
So, should we put a hold on this until the docker gcc image and Mac gcc get sorted? Should we use a workaround for people eager to get this up and going? What are you thoughts, @msarahan?

**Comment by msarahan** [comment]
Go ahead and add libgcc from defaults for now.  That will get a new version
on Linux, but that's fine for now.  Is the current docker image gcc 4.8?
If so, let's roll with it for this recipe.  For Mac, that's also a
discussion that could drag on.  No sense holding this up for it.

I want to get my docker images into a more official place than my personal
repo before anyone depends on them.  I'm working on getting them up at the
continuum account on docker hub. Is that OK with you guys (@pelson @ocefpaf
@jjhelmus @jakirkham), or would you rather build and host conda forge
versions?  The docker recipes will always be available, but I haven't
figured out if I'll be able to build gcc with the automated build.  The
most important thing to me is that we avoid multiple docker files floating
around, so that our system stays together.

**Comment by jakirkham** [comment]
> Is that OK with you guys (@pelson @ocefpaf @jjhelmus @jakirkham), or would you rather build and host conda forge versions?

The big thing, at least to me, is that we have a way to access it. That way, if something needs fixing, we can start working on it ASAP. Though, of course, as this is super essential to everything we are doing here. It makes sense to do CI, some testing, and go through an appropriate review. This might give us a lead on a hypothetical problem we might be dealing with here. For instance, reverting a troublesome patch and getting the container rebuilt ASAP until we can identify the problem adequately and fix it.

> The docker recipes will always be available, but I haven't figured out if I'll be able to build gcc with the automated build.

Based on what you said last time, I think this is one of the trickier issues. Once appropriate infrastructure can be configured, it will be a bit easier to see where this lives and how we might best interact with it.

For Jupyter's docker stacks, we ssh on to a machine for all our builds and only rebuild after merging a PR. Not elegant, but it is effective at addressing this problem.

> The most important thing to me is that we avoid multiple docker files floating around, so that our system stays together.

Agreed. Multiple docker files would be a mess. This is part of the reason that I think having a way for us to interact with the docker file and rebuild it is essential. Else we find ourselves in a situation where we need to fork it and rebuild it most likely to solve a temporary problem.

**Comment by jakirkham** [comment]
@ocefpaf, any straggling commits you want to make or are you ready to see this merged? :smile:

**Comment by ocefpaf** [comment]
> @ocefpaf, any straggling commits you want to make or are you ready to see this merged? :smile:

Happy to see this merged :grin: 

(Also happy with the discussion. I guess I am starting to understand the state of things now. Thanks @msarahan and @jakirkham.)

**Comment by ocefpaf** [comment]
Thanks @jakirkham :tada:

**Comment by jakirkham** [comment]
Thanks @ocefpaf! I really appreciate your patience while we are trying to sort this stuff out.

**Comment by ocefpaf** [comment]
> Thanks @ocefpaf! I really appreciate your patience while we are trying to sort this stuff out.

Conda-forge's goal is to provide high-quality, reproducible, peer reviewed  packages. If I had to wait an year for that I would be OK :wink:
